### PR TITLE
feat(ledger): introduce hardfork combinator shapes and refactor ledger

### PR DESCRIPTION
**Pull request by chrisguiney**
Relates to #1857 

---
## Summary by cubic
Introduce a new `ledger/hardfork` module with HFC shapes, summaries, and transition handling. Slot/time/epoch conversions and era history now use it, fixing cross‑era mappings and enforcing safe‑zone and stability‑window rules.

- **New Features**
  - Added `ledger/hardfork` with `Shape`, `Summary`, `EraParams`, `TransitionInfo`/`TransitionState`, and `BuildSummary`, with validation and tests.
  - Added `eras.StabilityWindowForEra` and `LedgerState.HardForkSummary` to compute era bounds from the epoch cache, tip, current era, and transition info.

- **Refactors**
  - Routed `SlotToTime`, `TimeToSlot`, `SlotToEpoch`, and `queryHardforkEraHistory` through `hardfork.Summary` for boundary‑safe cross‑era behavior and correct projection past tip.
  - Moved all transition types and references to `hardfork`; tightened error handling for before‑genesis and malformed era params.

<sup>Written for commit 5a51015e383005730bcafb2820e5af90fefbcb9b. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Multi-era hardfork support: accurate time/slot/epoch conversions, era summaries, and era lookups.
  * Exposed transition states (Unknown/Known/Impossible) with clearer transition reporting.

* **Refactoring**
  * Unified shape-based hardfork model and consolidated transition handling; ledger now builds and uses hardfork summaries.

* **Bug Fixes**
  * Safer validations and edge-case handling for slot/time conversions and malformed genesis parameters.

* **Tests**
  * Extensive regression and cross-era test coverage added.

### CODE REVIEW

File: ledger/hardfork/summary_test.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+				Start:  hardfork.Bound{},
+				End:    &end,
+				Params: byronParams,
+			},
+			{
+				EraID:  6,
+				Start:  end,
+				End:    nil, // unbounded
+				Params: shelleyParams,
+			},
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ledger/hardfork/summary_test.go:69]: not fixing

File: ledger/hardfork_summary_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	ls := &LedgerState{
+		epochCache: []models.Epoch{
+			{EpochId: 0, StartSlot: 0, SlotLength: 1000, LengthInSlots: 100, EraId: 1},
+		},
+		config: LedgerStateConfig{CardanoNodeConfig: &cardano.CardanoNodeConfig{}},
+	}
+	sum, err := ls.HardForkSummary()
+	require.NoError(t, err)
+	assert.True(t, sum.SystemStart.IsZero(),
+		"missing shelley genesis ⇒ SystemStart is zero time")
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ledger/hardfork_summary_test.go]: fixed

File: ledger/hardfork/summary.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+// SlotToTime converts an absolute slot number to its wall-clock start time.
+func (s *Summary) SlotToTime(slot uint64) (time.Time, error) {
+	era, err := s.eraForSlot(slot)
+	if err != nil {
+		return time.Time{}, err
+	}
+	slotsIntoEra := slot - era.Start.Slot
+	rel := era.Start.RelativeTime + time.Duration(slotsIntoEra)*era.Params.SlotLength
+	return s.SystemStart.Add(rel), nil
```
[the remark this answers is not in the document]
Comment (chrisguiney) [reply ledger/hardfork/summary.go:150]: addressed by annotating w/ `#nosec G115`

**Review by wolf31o2 — APPROVED** [review]
